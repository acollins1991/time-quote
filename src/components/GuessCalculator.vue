<template>
    <div class="text-5xl mt-5">
        <span class="minutes">{{ minutes }}m </span>
        <span class="hours">{{ hours }}h </span>
        <span class="days">{{ days }}d </span>
    </div>
</template>

<script>
export default {
    name: 'GuessCalculator',
    data(){
        return {
            minutes: 0,
            hours: 0,
            days: 0,
            totalMinutes: 0
        }
    },
    methods: {
        calculateEstimation: function(event){
            const bestGuess = event.bestGuess
            const optimisticGuess = event.optimisticGuess
            const pessimisticGuess = event.pessimisticGuess

            // PERT calc
            const totalMinutes = ( optimisticGuess + ( bestGuess*4 ) + pessimisticGuess ) / 6

            this.totalMinutes = totalMinutes

            // below 1 hour
            if( totalMinutes <= 59 ) {
                this.minutes = Math.round(totalMinutes)
            } 
            // at least 1 hour but less than 1 day
            else if( totalMinutes >= 60 && totalMinutes <= 1439 ) {
                var hours = totalMinutes / 60
                var rhours = Math.floor(hours)
                var minutes = (hours - rhours) * 60
                var rminutes = Math.round(minutes)

                this.minutes = rminutes 
                this.hours = rhours
            }
            // at 1 day
            else if( totalMinutes >= 1440 ) {
                var days = ( totalMinutes / 60 ) / 24
                var rdays = Math.floor(days)
                var hours = (days - rdays) * 24
                var rhours = Math.floor(hours)
                var minutes = (hours - rhours) * 60
                var rminutes = Math.round(minutes)

                this.minutes = rminutes 
                this.hours = rhours
                this.days = rdays
            }

        }
    },
    mounted(){
        this.$root.$on('guessesSubmitted', this.calculateEstimation)
    }
}
</script>
