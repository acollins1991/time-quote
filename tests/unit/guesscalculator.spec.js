import { shallowMount, createWrapper } from '@vue/test-utils'
import GuessCalculator from '@/components/GuessCalculator.vue'

describe( 'GuessCalculator.vue', () => {
    it('Renders initial values, 0m 0h 0d', () => {
        const wrapper = shallowMount(GuessCalculator)
        expect(wrapper.find('span.minutes').text()).toEqual('0m')
        expect(wrapper.find('span.hours').text()).toEqual('0h')
        expect(wrapper.find('span.days').text()).toEqual('0d')
    })
    it('Calculates the estimation on guessSubmitted root emitted event using PERT formula, 5mins', () => {

        const wrapper = shallowMount(GuessCalculator)
        const rootWrapper = createWrapper(wrapper.vm.$root)

        rootWrapper.vm.$emit('guessesSubmitted', { 'bestGuess': 5, 'optimisticGuess': 5, 'pessimisticGuess': 5 })

        expect(rootWrapper.emitted().guessesSubmitted).toBeTruthy()

        // 1505mins converts to 1day 1hours 5mins
        expect(wrapper.vm.$data.minutes).toEqual(5)
        expect(wrapper.vm.$data.hours).toEqual(0)
        expect(wrapper.vm.$data.days).toEqual(0)

    })
    it('Calculates the estimation on guessSubmitted root emitted event using PERT formula, 1439mins', () => {

        const wrapper = shallowMount(GuessCalculator)
        const rootWrapper = createWrapper(wrapper.vm.$root)

        rootWrapper.vm.$emit('guessesSubmitted', { 'bestGuess': 1439, 'optimisticGuess': 1439, 'pessimisticGuess': 1439 })

        expect(rootWrapper.emitted().guessesSubmitted).toBeTruthy()

        // 1505mins converts to 1day 1hours 5mins
        expect(wrapper.vm.$data.minutes).toEqual(59)
        expect(wrapper.vm.$data.hours).toEqual(23)
        expect(wrapper.vm.$data.days).toEqual(0)

    })
    it('Calculates the estimation on guessSubmitted root emitted event using PERT formula, 1505mins', () => {

        const wrapper = shallowMount(GuessCalculator)
        const rootWrapper = createWrapper(wrapper.vm.$root)

        rootWrapper.vm.$emit('guessesSubmitted', { 'bestGuess': 1505, 'optimisticGuess': 1505, 'pessimisticGuess': 1505 })

        expect(rootWrapper.emitted().guessesSubmitted).toBeTruthy()

        // 1505mins converts to 1day 1hours 5mins
        expect(wrapper.vm.$data.minutes).toEqual(5)
        expect(wrapper.vm.$data.hours).toEqual(1)
        expect(wrapper.vm.$data.days).toEqual(1)

    })
})