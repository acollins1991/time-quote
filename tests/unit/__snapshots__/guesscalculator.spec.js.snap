// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GuessCalculator.vue renders correctly 1`] = `
<div
  class="text-5xl mt-5"
>
  <span
    class="days"
  >
    0d 
  </span>
   
  <span
    class="hours"
  >
    0h 
  </span>
   
  <span
    class="minutes"
  >
    0m 
  </span>
</div>
`;
